airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/AirbyteType.kt

```diff
@@ -0,0 +1,157 @@
+package io.airbyte.cdk.message
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+sealed class AirbyteType {
+    companion object {
+        fun fromAirbyteJsonSchema(jsonSchema: ObjectNode): ObjectType {
+            val airbyteSchema = fromJsonSchema(jsonSchema)
+            if (airbyteSchema !is ObjectType) {
+                throw IllegalArgumentException("Top-level schema must be an object (got $airbyteSchema)")
+            }
+            return airbyteSchema
+        }
+
+        fun fromJsonSchema(schema: ObjectNode): AirbyteType {
+            if (schema.has("type")) {
+                val schemaType = schema.get("type")
+                if (schemaType.isTextual) {
+                    return when (schema.get("type").asText()) {
+                        "string" -> StringType.fromJsonSchema(schema)
+                        "boolean" -> BooleanType
+                        "integer" -> IntegerType
+                        "number" -> NumberType.fromJsonSchema(schema)
+                        "array" -> ArrayType.fromJsonSchema(schema)
+                        "object" -> ObjectType.fromJsonSchema(schema)
+                        "null" -> NullType
+                        else -> throw IllegalArgumentException(
+                            "Unknown type: ${
+                                schema.get("type").asText()
+                            }"
+                        )
+                    }
+                } else if (schemaType.isArray) {
+                    return UnionType.fromJsonSchemaTypeArray(schemaType as ArrayNode, schema)
+                }
+            } else {
+                val options = schema.get("oneOf") ?: schema.get("anyOf") ?: schema.get("allOf")
+                if (options != null) {
+                    return UnionType.fromUnionOptions(
+                        options.map { fromJsonSchema(it as ObjectNode) }
+                    )
+                }
+            }
+
+            throw IllegalArgumentException("Unknown schema type: $schema")
```

Comment:
Second point first: I'd rather have distinct types for things like un-schematized objects and arrays, because otherwise every `when` ends up making the same checks to distinguish types, which seems error-prone, whereas if you don't care about the distinction, it's more natural to do this than the reverse:
```
when (abt) {
  Object,
  ObjectWithoutSchema,
  ObjectWithEmptySchema ->
}
```

For the first point, `ObjectWithEmptySchema` can come out of the top in that case. For now I'll relax the enforcement that the schema be a schematized `Object` type.

Reply:
agree that ObjectWithoutSchema deserves its own class for sure. By definition, it doesn't fit into the `Object(properties: Map)` struct. (we _could_ do it as `properties: Map?`... probably a bad idea though)

I forget - what do we currently do with a field with schema `{type: object, properties: {}}`? i.e. do we actually generate an empty avro struct? or do we stringify it?  (if we stringify, then yes, agree that ObjectWithEmptySchema deserves its own class.)

I don't see what's special about ArrayWithoutSchema though - IMO we should treat these three schemas identically?
* `{type: array}` (no explicit schema)
* `{type: array, items: {}}` (explicit accept-all schema)
* `{type: array, items: {type: potato}}` (unrecognized schema)

(underlying assumption: destinations convert Unknown schema to "widest type". e.g. stringify json, variant, etc.)

Reply:
Currently we call it OBJECT_WITHOUT_PROPERTIES and stringify it. (AFAIK, this just occurs with a lot of API sources that don't report their schemas, so the intended meaning is "no schema" not "no fields".)